### DISCUSSION TITLE
My Homebox "meta feedback" after adding 3,000+ items

### DISCUSSION THREAD

**Discussion by danielrosehill** [hay-kot/homebox#862] [General]
Overall sentiment:

I have almost nothing but praise. 

I've been intuitively looking for something like Homebox for years but took a quick look into some open source ERPs and thought "these are WAY too complicated!"

Homebox would have saved me so much time and also money (I got to the point at which I was rebuying basic items like audio adapters because it was quicker than rummaging through loads of boxes).

I'm also seeing huge value in Homebox as a decluttering tool! 

Every time I come across an item that I don't need any more but which is in good condition I just add a "to giveaway" tool. I can then share the pics on FB / directly with people / whatever:

![giveaway](https://github.com/hay-kot/homebox/assets/40233203/fa0726ab-2f91-4e5e-b6ee-8e486e1ede22)

The UI is logical although I think some basic user guidance on what the suggested hierarchy is would be helpful (should boxes be assets and their contents child assets or should boxes be locations? I shifted halfway through to the former approach but I think both have their merits). It's also super quick.

**The following are the things I would highlight as the most pressing impediments in the UI at the time I'm writing this:**

-  1: The refresh thing. I can't remember the details of the upstream bug. But it's currently VERY easy to accidentally overwrite data if you forget to refresh a page. 

- 2: The search. I don't know how the search is configured, but the parameters are a little too exacting IMO. Frequently I don't get results because of little things like I missed a capital letter or a hyphen. If there's a specificity parameter perhaps it could be turned down a little?

- 3: Multi image support. It would be enormously helpful to be able to upload multiple images at once. I find adding photos super helpful (add warranty stickers, SN stickers, etc). It would also be nice to be able to jump between a "gallery" with the arrow keys on the keyboard. 

- 4: I'm guessing few people are using this with a standalone barcode scanner, but a barcode reader page would be super cool. Those typically work by expecting the input from the scanner and jumping too it. 

Final thought: I don't know if the project has any commercial ambitions, but I reckon people would love this as a SaaS tool. I've checked out the alternatives and wasn't wowed.

Hope that might be of some utility!

**Comment by ScottEgan** [comment]
> "...(should boxes be assets and their contents child assets or should boxes be locations? I shifted halfway through to the former approach but I think both have their merits)."

@danielrosehill Do you think you could expand on your findings between the two organization styles? I'm starting down a similar path and would like hear about what you've learnt so far? 

What I've found is that if a box is an asset uploading via a CSV doesn't work as well as you have to manually assign the child assets.

**Reply by danielrosehill** [reply]
@ScottEgan 

Absolutely!

The simple answer is that I would highly recommend creating your boxes as assets rather than locations.

I spent an afternoon I'd rather forget peeling off a bunch of labels and recreating all 70 or so boxes one by one!

If you configure the boxes as "locations":

1: They're not searchable in the system, so it's impossible to easily see which "box" you're up to if you're numbering them sequentially (this created a whole bunch of problems for me including accidentally creating duplicates).

2: Halfway through my inventory a storage unit we bought sort of came apart. This made me realise that it's actually sensible to regard the storage units themselves as assets in case you might need to store warranty documents (again thinking of items like sheds rather than cheap plastic boxes)

3: When all your boxes look very similar, it's nice to be able to add photos to identify them. You can only do that with assets.

So what I've done:

1: Create an asset for each box
2: All its contents are sub-assets

The only negative is that the breadcrumb nav doesn't let you see the full nesting chain if you're putting the boxes (assets) as subassets of storage systems but..... that's my complicated setup.

Some pics:

![image](https://github.com/hay-kot/homebox/assets/40233203/8d33442b-eb46-4420-8e31-1aa9f7487d0f)

![image](https://github.com/hay-kot/homebox/assets/40233203/c8cd21b2-7718-4eb0-932e-659208f61007)

![image](https://github.com/hay-kot/homebox/assets/40233203/3b1fa7b2-7b9b-4861-a4ae-5b8f11f0026b)

![image](https://github.com/hay-kot/homebox/assets/40233203/242e0bc4-cd08-4d15-9662-a76668d26e29)

![image](https://github.com/hay-kot/homebox/assets/40233203/01d5bdcd-d925-4bce-8947-6120aeb7b642)

![image](https://github.com/hay-kot/homebox/assets/40233203/4a09f886-0511-463a-b23a-24c469b2ff70)

**Reply by ScottEgan** [reply]
Thank you! Very informative! 

If I may ask a couple more questions:

1. If your shelves are also assets, then when you search for an item, you can then work your way back up the tree to figure out what shelf it is on and what location that shelf is in?
2. Is there a quick way to assign children to an asset or do they have to be done one by one?

**Reply by danielrosehill** [reply]
1. Yes, exactly.

So for example I can search for this load balancing router:

![image](https://github.com/hay-kot/homebox/assets/40233203/fe5fdc74-79b4-4cd1-be7b-7fc9393e82ea)

It's in box 74 (!). And when I click on the box number, I've mapped that onto an individual shelf:

![image](https://github.com/hay-kot/homebox/assets/40233203/3775641e-8c8b-4971-98e8-a44dd9330503)

In this case it's shelving (asset) -> shelves (child assets, level 1) -> box (child asset, level 2) -> item (child asset, level 3) ... or something like that.

2. No the UI and workflow needs improvement in this regard. I have to create them and then assign individually.

**Reply by ronbruins** [reply]
I finally was able to create a parentID for an asset through the API. So this helps me a lot as now I have the following script:

- pictures in folder
Folder is the name of the parent
Pics have name of the asset

Python os.walk crawls through the folder and creates item and returns id. Then it creates item based on photo. The returned id is then used to add it to the parent asset (which id is known)

This automatically creates item/asset and all sub assets and uploads the photo to each one. 

So all I need to do is take pics and rename them accordingly. 

I'm gonna enhance this with labels and descriptions,
Maybe via a text file or something, but that's WIP

**Reply by ScottEgan** [reply]
> I finally was able to create a parentID for an asset through the API. So this helps me a lot as now I have the following script:
> 
> - pictures in folder
> Folder is the name of the parent
> Pics have name of the asset
> 
> Python os.walk crawls through the folder and creates item and returns id. Then it creates item based on photo. The returned id is then used to add it to the parent asset (which id is known)
> 
> This automatically creates item/asset and all sub assets and uploads the photo to each one. 
> 
> So all I need to do is take pics and rename them accordingly. 
> 
> I'm gonna enhance this with labels and descriptions,
> Maybe via a text file or something, but that's WIP

Would love to see the code if you don't mind sharing.

**Reply by ronbruins** [reply]
I just tested it today within postman and some python code with hard coded ids. When I incorporated it in my original script I don’t mind sharing it. 

I can already share the main script. Which does exactly the same but it creates a location with assets instead of asset with sub assets. 

It's also on discord as a reply to Daniel.

**Reply by ronbruins** [reply]
```
import requests
import json
import os

baseurl = "http://<hb_IP>:3100/api/v1/"
user = "user%40domain.com"
password = "password"
loc_folder = "/absolute_path/locations"

def main():
    token = login()
    get_file_list(token)

def get_file_list(token):
    for root, dirs, files in os.walk(loc_folder):
        for location in dirs:
            print(f"Location: {location}")
            location_id = create_location(token,location)
            parent_itemname = f"parent_{location}"
            parent_item_id = create_item(token,location_id,parent_itemname)
            item_location = f"{loc_folder}/{location}"
            for root, dirs, files in os.walk(item_location):
                for item in files:
                    itemname = item.split('.')
                    itemname = itemname[0]
                    print(f"itemname: {itemname}")
                    print(f"Item: {item}")
                    item_id = create_item(token,location_id,itemname)
                    upload_photo(token,item_id,item,item_location)
                    update_item(token,parent_item_id,item_id,location_id,itemname)

def hb_post(token,method,api,json_data):
    url = baseurl + api
    headers = {
        'Authorization': f'{token}',
        'accept': 'application/json',
        'Content-Type': 'application/json',
    }
    if "attachments" in api:
        headers = {
        'accept': 'application/json',
        'Authorization': f'{token}',
        }
        response = requests.request(method, url, headers=headers, files=json_data)
    else:
        response = requests.request(method,url, headers=headers, json=json_data)
        item = json.loads(response.text)
        item_id = item['id']
        return item_id


def update_item(token,parent_item_id,item_id,location_id,name):
    api = f"items/{item_id}"
    method = "PUT"
    json_data = {
        "id":f"{item_id}",
        "name":f"{name}",
        "locationId":f"{location_id}",
        "parentId":f"{parent_item_id}"
        }
    item_id = hb_post(token,method,api,json_data)
    return item_id

def create_item(token,location_id,name):
    api = "items"
    method = "POST"
    json_data = {
        'description': None,
        'labelIds': [],
        'locationId': f'{location_id}',
        'name': f'{name}',
        'parentId': None,
    }
    item_id = hb_post(token,method,api,json_data)
    return item_id

def create_location(token,location):
    api = "locations"
    method = "POST"
    json_data = {
        "description": None,
        "name": f"{location}",
        "parentId": None
        }
    location_id = hb_post(token,method,api,json_data)
    return location_id

def upload_photo(token,item_id,item,item_location):
    api = f"items/{item_id}/attachments"
    fp = f"{item_location}/{item}"
    method = "POST"
    files_data = {
        'file': (f'{item}', open(fp, 'rb'), 'image/jpeg'),
        'type': (None, 'photo'),
        'name': (None, f'{item}'),
    }
    upload_post = hb_post(token,method,api,files_data)

def login():
    api = "users/login"
    url = baseurl + api
    payload = f'username={user}&password={password}'
    headers = {
        'Accept': 'application/json',
        'Content-Type': 'application/x-www-form-urlencoded',
    }
    response = requests.request("POST", url, headers=headers, data=payload)
    hbresponse = json.loads(response.text)
    token = hbresponse['token']
    print(token)
    return token

if __name__ == '__main__':
    main()
```

**Reply by ronbruins** [reply]
Above the initial version of creating a location, asset and subsets from a folder with structure:

```
./locations/AssetTestPoetsBox
./locations/AssetTestPoetsBox/Nanotech_Impregneer_2 Medium.jpeg
./locations/AssetTestPoetsBox/AutoShampoo Medium.jpeg
./locations/AssetTestPoetsBox/SchoenBorstel Medium.jpeg
./locations/AssetTestPoetsBox/WaterStop_Schoen_Neutraal Medium.jpeg
./locations/AssetTestPoetsBox/SuedeBorsteltje Medium.jpeg
.....
```

It now creates a parent item based on the location name, this needs to be changed to allow dynamic naming. However I also need to included a check for the 'location' itself if it exists use it, if not create it. Still work in progress, but because requested here the quick & dirty version. I will share my repo with this when more mature :-)
